bdai_ros2_wrappers/bdai_ros2_wrappers/action.py

```diff
@@ -0,0 +1,583 @@
+        """Callback on action result future resolution."""
+        exception = result_future.exception()
+        if exception is None:
+            self._finalization_future.set_result(result_future.result().status)
+        else:
+            self._finalization_future.set_exception(exception)
+        if self._feedback_tape is not None:
+            self._feedback_tape.close()
+
+    @property
+    def acknowledgement(self) -> Future:
+        """Get future to wait for action acknowledgement (either acceptance or rejection)."""
+        return self._acknowledgement_future
+
+    @property
+    def acknowledged(self) -> bool:
+        """Check if action was acknowledged (i.e. accepted or rejected)"""
+        return self._acknowledgement_future.done()
+
+    @property
+    def goal_handle(self) -> ClientGoalHandle:
+        """
+        Get action goal handle.
+
+        Raises:
+            RuntimeError: if action has not been accepted or rejected yet.
+        """
+        if not self._goal_handle_future.done():
+            raise RuntimeError("Action not acknowledged yet")
+        return self._goal_handle_future.result()
+
+    @property
+    def finalization(self) -> Future:
+        """
+        Get future to wait for action finalization.
+
+        Action rejection also counts as finalization.
+        """
+        return self._finalization_future
+
+    @property
+    def finalized(self) -> bool:
+        """Check if action was finalized."""
+        return self._finalization_future.done()
+
+    @property
+    def result(self) -> Any:
+        """
+        Get action result.
+
+        Raises:
+            RuntimeError: if action is still executing.
+        """
+        if self._result_future is None or not self._result_future.done():
+            raise RuntimeError("Action still executing")
+        return self._result_future.result().result
+
+    @property
+    def tracks_feedback(self) -> bool:
+        """Check if action future tracks action feedback."""
+        return self._feedback_tape is not None
+
+    @property
+    def feedback(self) -> Sequence:
+        """
+        Get all buffered action feedback.
+
+        Action must have been accepted before any feedback can be fetched.
+
+        Raises:
+            RuntimeError: if action feedback tracking is not enabled,
+            action acknowledgement has not been received yet, or
+            action was not accepted.
+        """
+        if self._feedback_tape is None:
+            raise RuntimeError("Action feedback tracking is disabled")
+        if not self._goal_handle_future.done():
+            raise RuntimeError("Action not acknowledged yet")
+        goal_handle = self._goal_handle_future.result()
+        if not goal_handle.accepted:
+            raise RuntimeError("Action not accepted")
+        return list(self._feedback_tape.content())
+
+    def feedback_stream(self, *, buffer_size: Optional[int] = None, timeout_sec: Optional[float] = None) -> Iterable:
```

Comment:
Is the `buffer_size` used? 

Also is there a way to start the stream with the next received feedback?

Reply:
> Is the `buffer_size` used?

Ah, good catch.

> Also is there a way to start the stream with the next received feedback?

Ha, I had that feature at some point and eventually removed it as I figured no one would need it. We can add it back.

Reply:
Done and feature restored in https://github.com/bdaiinstitute/ros_utilities/pull/58/commits/329efc3935f1f9fdc187d6157e4d218f4289adfd.